tidal-link/src/hs/Sound/Tidal/Clock.hs

```diff
@@ -0,0 +1,341 @@
+module Sound.Tidal.Clock where
+
+import qualified Sound.Tidal.Link as Link
+import qualified Sound.Osc.Fd as O
+
+import Control.Concurrent
+import Control.Monad         (when)
+import Control.Monad.Reader  (ReaderT, runReaderT, ask)
+import Control.Monad.State   (StateT, liftIO, evalStateT, get, put, modify)
+
+import Foreign.C.Types (CDouble (..))
+import Data.Int (Int64)
+import Data.Coerce (coerce)
+import System.IO (hPutStrLn, stderr)
+
+type Time = Rational
+
+-- | representation of a tick based clock
+type Clock
+  = ReaderT ClockMemory (StateT ClockState IO)
+
+-- | internal read-only memory of the clock
+data ClockMemory
+  = ClockMemory
+  {clockConfig :: ClockConfig
+  ,clockRef :: ClockRef
+  }
+
+-- | internal mutable state of the clock
+data ClockState
+  = ClockState
+  {ticks  :: Int64
+  ,start  :: Link.Micros
+  ,nowArc :: (Time, Time)
+  ,nudged :: Double
+  } deriving Show
+
+-- | reference to interact with the clock, while it is running
+data ClockRef
+  = ClockRef
+  {rAction :: MVar ClockAction
```

Comment:
but you can do exactly what you wrote, no need for a list, the only thing is that you can't do them *simultaneously*, i.e. in the same tick (which you could theoretically do with the list implementation, but is this really neccessary?)

Reply:
As an example, I think the entire contents of a `BootTidal` file would generally be executed on a single tick, and it seems like an unintuitive limitation that you can't set both nudge and cps in your boot file.

Is there some benefit to not executing these commands in the same tick?

Reply:
i just think the feature was never used anywhere and the only reason it was a list before was basically for having the empty list as a noop

in the bootfile this will actually be the case since it just starts the clock with a given config, no clock actions are executed at that time.

to make use of this in terms of the API we would have to introduce functions like `setCPSandCycle` etc., i'm not sure if that's very helpful, no?

Reply:
You don't need a new function, you just need a regular `do` block. I just tried evaluating the following code in your PR, and the cps doesn't change:
```haskell
do streamSetCPS tidal 0.2
   setCycle 0
```

If you call the equivalent code in `1.9-dev`, then the cps does change.

Reply:
That said, I don't fully understand the context that led to the creation of the `TempoAction` structure in the first place, so maybe I'm not thinking about this correctly. I'll set this aside and move back into general discussion. :smiley:

Reply:
hmm well there is no `streamSetCPS` on 1.9-dev, there is `setcps` which is `once . cps` i think.
but i would expect your code to execute both actions.. i'll have to look into it

Reply:
ok, thinking about this more:
```
do streamSetCPS tidal 0.2
   setCycle 0
```
here the problem is that the same MVar is modified twice, before it has the clock thread has the chance to accsess the first modification. I can implement a simple solution by using `TVars` from` Control.Concurrent.STM` instead, which are apperently less efficient but allow for more guarantees. I think in this case this is a good trade off, since these actions are not used so much anyways.